### PR TITLE
feat(parser): add models for API GW Websockets events

### LINKED ISSUE #5543: Feature request: Parser models for API GW Websockets Events
### Use case

I can implement it myself, it's been a while :)
I'd like to present/demo this feature in an upcoming blog post

3 times of events:
connect: https://gist.github.com/ran-isenberg/7c73a1db17a5d47ac5da5e19f0a3529f

message over a route : https://gist.github.com/ran-isenberg/35ceae2179ed9c26e459450825561369

disconnect:  https://gist.github.com/ran-isenberg/2bffedaacc4423d7d23901b803cc280a

### Solution/User Experience

Add Parser models and envelope to parse the event.
Same experience as API GW usage.

### Alternative solutions

```markdown
write the model by myself and use pure pydantic like a savage.
```

### Acknowledgment

- [x] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [x] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by ran-isenberg**
**Issue number**: #5543

## Summary
Added support for parser and websockers api
### Changes

- Parser envelope for the message type.
- models for message, connect and disconnect
- unit tests
- 3 input files
- documentation

Used this doc and my own files. 
https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api-integration-requests.html

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @ran-isenberg! Thanks for submitting this PR!

Assigning @anafalcao to review this.

**Comment by anafalcao** [comment]
Thank you for submitting this PR so promptly, @ran-isenberg ! I was off today, but tomorrow I'll begin to review it :)

File: aws_lambda_powertools/utilities/parser/models/__init__.py [hunk 1]
```diff
@@ -7,6 +7,16 @@
...
 )
+from .apigw_websocket_api import (
+    APIGatewayWebSocketApiConnectEventModel,
+    APIGatewayWebSocketApiConnectEventRequestContext,
+    APIGatewayWebSocketApiDisconnectEventModel,
+    APIGatewayWebSocketApiDisconnectEventRequestContext,
+    APIGatewayWebSocketApiEventIdentity,
+    APIGatewayWebSocketApiEventRequestContextBase,
+    APIGatewayWebSocketApiMessageEventModel,
+    APIGatewayWebSocketApiMessageEventRequestContext,
```
Comment (anafalcao) [inline aws_lambda_powertools/utilities/parser/models/__init__.py]:
```suggestion
    APIGatewayWebSocketConnectEventModel,
    APIGatewayWebSocketConnectEventRequestContext,
    APIGatewayWebSocketDisconnectEventModel,
    APIGatewayWebSocketDisconnectEventRequestContext,
    APIGatewayWebSocketEventIdentity,
    APIGatewayWebSocketEventRequestContextBase,
    APIGatewayWebSocketMessageEventModel,
    APIGatewayWebSocketMessageEventRequestContext,
```

Comment (anafalcao) [reply aws_lambda_powertools/utilities/parser/models/__init__.py]: Hey @ran-isenberg ! I think we can remove Api from the names

File: aws_lambda_powertools/utilities/parser/models/__init__.py [hunk 2]
```diff
@@ -105,6 +115,14 @@
...
     "APIGatewayProxyEventV2Model",
     "ApiGatewayAuthorizerRequestV2",
+    "APIGatewayWebSocketApiEventIdentity",
+    "APIGatewayWebSocketApiMessageEventModel",
+    "APIGatewayWebSocketApiMessageEventRequestContext",
+    "APIGatewayWebSocketApiConnectEventModel",
+    "APIGatewayWebSocketApiConnectEventRequestContext",
+    "APIGatewayWebSocketApiDisconnectEventRequestContext",
+    "APIGatewayWebSocketApiDisconnectEventModel",
+    "APIGatewayWebSocketApiEventRequestContextBase",
```
Comment (anafalcao) [inline aws_lambda_powertools/utilities/parser/models/__init__.py]: same here

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket_api.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from typing import Dict, List, Literal, Optional, Type, Union
+
+from pydantic import BaseModel
+from pydantic.networks import IPvAnyNetwork
+
+
+class APIGatewayWebSocketApiEventIdentity(BaseModel):
```
Comment (anafalcao) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket_api.py]: same here (for the Api name) and other similars

**Review by anafalcao — CHANGES_REQUESTED** [review]
Hey @ran-isenberg, I suggested some changes in a initial review. 
I see the CI is failing, but don't worry, it's not related to your PR. We are verifying it

**Comment by ran-isenberg** [comment]
@anafalcao  all fixed! also renamed one of the files and found a documentation mistake.
Please review again :)

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from typing import Dict, List, Literal, Optional, Type, Union
+
+from pydantic import BaseModel
+from pydantic.networks import IPvAnyNetwork
+
+
+class APIGatewayWebSocketEventIdentity(BaseModel):
```
Comment (anafalcao) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket.py:8]: In my tests I see the userAgent always come inside Indentity:

`'identity': {'userAgent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36', 'sourceIp': '179.164.64.249'}`

Can you please check if it's Optional or not?

File: docs/utilities/parser.md [hunk 5]
```diff
@@ -188,8 +191,9 @@ You can use pre-built envelopes provided by the Parser to extract and parse spec
 | **KinesisFirehoseEnvelope**   | 1. Parses data using `KinesisFirehoseModel` which will base64 decode it. ``2. Parses records in in` Records` key using your model`` and returns them in a list.                                          | `List[Model]`                      |
 | **SnsEnvelope**               | 1. Parses data using `SnsModel`. ``2. Parses records in `body` key using your model`` and return them in a list.                                                                                         | `List[Model]`                      |
 | **SnsSqsEnvelope**            | 1. Parses data using `SqsModel`. `` 2. Parses SNS records in `body` key using `SnsNotificationModel`. `` 3. Parses data in `Message` key using your model and return them in a list.                  | `List[Model]`                      |
+| **ApiGatewayV2Envelope**      | 1. Parses data using `APIGatewayWebSocketApiMessageEventModel`. ``2. Parses `body` key using your model`` and returns it.                                                                                            | `Model`                            |
```
Comment (anafalcao) [inline docs/utilities/parser.md]:
```suggestion
| **ApiGatewayV2Envelope**      | 1. Parses data using `APIGatewayProxyEventV2Model`. ``2. Parses `body` key using your model`` and returns it.                                                                                            | `Model`                            |
```
Please, keep as is

**Review by anafalcao — DISMISSED** [review]
Hey @ran-isenberg ! Looking good so far, but there are a few comments. Can you take a look, please?

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py (hunk 4 above)
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parser/models/apigw_websocket.py:8]: Hey, I'm not near a computer in the next couple.of days. I attached the aws docs at tje start, could you verify it?

File: docs/utilities/parser.md (hunk 5 above)
Comment (ran-isenberg) [reply docs/utilities/parser.md]: Hey, I'm not near a computer in the next couple.of days Could you push the change?

**Comment by anafalcao** [comment]
Np, thanks again @ran-isenberg ! I pushed those small changes, but the overall looks very good! 
@leandrodamascena can you review it as well? 
Thanks!

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    headers: Dict[str, str]
+    multiValueHeaders: Dict[str, List[str]]
+    requestContext: APIGatewayWebSocketDisconnectEventRequestContext
+    isBase64Encoded: bool
+
+
+class APIGatewayWebSocketMessageEventModel(BaseModel):
+    requestContext: APIGatewayWebSocketMessageEventRequestContext
+    isBase64Encoded: bool
+    body: Optional[Union[str, Type[BaseModel]]] = None
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket.py]: nice one @ran-isenberg! Good point making the body accept a model to be serialized as the customer wants.

File: aws_lambda_powertools/utilities/parser/envelopes/apigw_websocket_api.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+from aws_lambda_powertools.utilities.parser.models import APIGatewayWebSocketMessageEventModel
+
+if TYPE_CHECKING:
+    from aws_lambda_powertools.utilities.parser.types import Model
+
+logger = logging.getLogger(__name__)
+
+
+class ApiGatewayWebSocketEnvelope(BaseEnvelope):
+    """API Gateway WebSockets API envelope to extract data within body key of messages routes
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/envelopes/apigw_websocket_api.py]:
```suggestion
    """API Gateway WebSockets envelope to extract data within body key of messages routes
```

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+from datetime import datetime
+from typing import Dict, List, Literal, Optional, Type, Union
+
+from pydantic import BaseModel, ConfigDict
+from pydantic.networks import IPvAnyNetwork
+
+
+class APIGatewayWebSocketEventIdentity(BaseModel):
+    model_config = ConfigDict(populate_by_name=True)
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket.py]: I understand why you added this here, but I think we can remove it @anafalcao

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    sourceIp: IPvAnyNetwork
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket.py]:
```suggestion
    source_ip: IPvAnyNetwork = Field(alias="sourceIp")
```

File: aws_lambda_powertools/utilities/parser/models/apigw_websocket.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    userAgent: Optional[str] = None
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/apigw_websocket.py]:
```suggestion
    user_agent: Optional[str] = = Field(None, alias="userAgent")
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @ran-isenberg! Thank you very much for sending this PR, it's very good as always.

I know that the event sent by AWS uses CamelCase by default and we can't change it for obvious reasons, but I would like to make the fields snake case on our side and work with Pydantic Fields. I also know that most of our events use CamelCase, but I would like to have this convention about capitalization - We've done this on newer models like Cloudformation Custom Resource, VPC Lattice 2, and Bedrock. Can you please change this @anafalcao? I left 2 examples that you can extend to any other.

After this change we are good to merge this PR! 🚀

File: aws_lambda_powertools/utilities/parser/envelopes/apigw_websocket_api.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/envelopes/apigw_websocket_api.py:1]: Please rename this file to `apigw_websocket.py`.

**Review by leandrodamascena — APPROVED** [review]
Thanks for addressing all the feedback @anafalcao! 

I'm approving this PullRequest and please open a new issue to add the same support for Event Source Data Classes!

APPROVED!!!

**Comment by ran-isenberg** [comment]
Whoohooo! Thx @anafalcao  and @leandrodamascena

**Pull request merged**